src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,200 @@
+
+    r_projs = map(0:lmax) do l
+        betas_l = filter(beta -> beta["angular_momentum"] == l, pseudo["beta_projectors"])
+        map(beta -> beta["radial_function"] ./ 2, betas_l)  # Ry -> Ha
+    end
+    
+    h = Matrix[]
+    count = 1
+    for l = 0:lmax
+        nproj_l = length(r_projs[l+1])
+        Dij_l = pseudo["D_ion"][count:count+nproj_l-1, count:count+nproj_l-1] .* 2  # 1/Ry -> 1/Ha
+        push!(h, Dij_l)
+        count += nproj_l
+    end
+
+    pswfcs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["radial_function"], pswfcs_l)
+    end
+    
+    pswfc_occs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["occupation"], pswfcs_l)
+    end
+    
+    rhoatom = pseudo["total_charge_density"]
+
+    return PspUpf(Zion, lmax, rgrid, drgrid, vloc, r_projs, h, pswfcs, pswfc_occs, rhoatom;
+                  identifier, description)
+end
+
+function PspUpf(Zion, lmax, rgrid::Vector{T}, drgrid, vloc, r_projs, h, pswfcs, pswfc_occs,
+                rhoatom; identifier="", description="") where {T <: Real}
+
+    vloc_interp = linear_interpolation((rgrid, ), vloc)
+    r_projs_interp = map(r_projs) do r_projs_l
+        map(r_projs_l) do r_proj  # Can't use views here; have to match `vloc_interp`'s type
+            ir_cut = lastindex(r_proj)
+            linear_interpolation((rgrid[1:ir_cut], ), r_proj)
+        end
+    end
+
+    r_vloc_corr_dr = (rgrid .* vloc .+ Zion) .* drgrid
+    r2_projs_dr = map(r_projs) do r_projs_l
+        @views map(r_projs_l) do r_proj
+            ir_cut = lastindex(r_proj)
+            rgrid[1:ir_cut] .* r_proj .* drgrid[1:ir_cut]
+        end
+    end
+
+    PspUpf{T,typeof(vloc_interp)}(Zion, lmax, rgrid, drgrid, vloc, r_projs, h, pswfcs,
+                                  pswfc_occs, rhoatom, vloc_interp, r_projs_interp,
+                                  r_vloc_corr_dr, r2_projs_dr, identifier, description)
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
+
+"""
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r<:Real)
+
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space
+distance r via linear interpolation on the real-space mesh.
+
+Note: UPFs store `r[j] β_{li}(r[j])`, so r=0 is undefined and will error.
+"""
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T)::T where {T <: Real}
+    psp.r_projs_interp[l+1][i](r) / r
+end
+
+"""
+    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q<:Real)
+
+For UPFs, the integral is transformed to the following sum:
+4π Σ{k} j_l(q r[k]) (r[k]^2 p_{il}(r[k]) dr[k])
+"""
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
+    s = zero(T)
+    @inbounds for ir = eachindex(psp.r2_projs_dr[l+1][i])
```

Comment:
Validate the l here or put the inbounds after indexing l, it is illegal otherwise (I just got a segfault from it)